pkg/auth/handler/login_auth_url_test.go

```diff
@@ -0,0 +1,172 @@
+package handler
+
+import (
+	"net/url"
+	"testing"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/sso"
+	"github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/server/skyerr"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestLoginAuthURLPayload(t *testing.T) {
+	Convey("Test LoginAuthURLRequestPayload", t, func() {
+		// callback URL and ux_mode is required
+		Convey("validate valid payload", func() {
+			payload := LoginAuthURLRequestPayload{
+				CallbackURL: "callbackURL",
+				UXMode:      sso.WebRedirect,
+			}
+			So(payload.Validate(), ShouldBeNil)
+		})
+
+		Convey("validate payload without callback url", func() {
+			payload := LoginAuthURLRequestPayload{
+				UXMode: sso.WebRedirect,
+			}
+			err := payload.Validate()
+			errResponse := err.(skyerr.Error)
+			So(errResponse.Code(), ShouldEqual, skyerr.InvalidArgument)
+		})
+
+		Convey("validate payload without UX mode", func() {
+			payload := LoginAuthURLRequestPayload{
+				CallbackURL: "callbackURL",
+			}
+			err := payload.Validate()
+			errResponse := err.(skyerr.Error)
+			So(errResponse.Code(), ShouldEqual, skyerr.InvalidArgument)
+		})
+	})
+}
+
+func TestSSOUXModeConvertor(t *testing.T) {
+	Convey("Test UXModeFromString", t, func() {
+		Convey("should convert string to sso.UXMode", func() {
+			modes := []sso.UXMode{
+				sso.WebRedirect,
+				sso.WebPopup,
+				sso.IOS,
+				sso.Android,
+			}
+			for _, m := range modes {
+				modeFromStr := sso.UXModeFromString(m.String())
+				So(modeFromStr, ShouldEqual, m)
+			}
+		})
+
+		Convey("should convert unknown string to sso.Undefined", func() {
+			modeFromStr := sso.UXModeFromString("some_string")
+			So(modeFromStr, ShouldEqual, sso.Undefined)
+		})
+	})
+}
+
+func TestLoginAuthURLHandler(t *testing.T) {
+	Convey("Test TestLoginAuthURLHandler", t, func() {
+		h := &LoginAuthURLHandler{}
```

Comment:
For the test case format, we had discussion before and thought JSON in/out format is clearer.

Our discussion: https://github.com/SkygearIO/skygear-server/pull/751

I will merge this PR after update `authURL` function, we can update this test later but please use JSON in/out format in the future PR.

Thanks @kaojohnny